src/solvers/smt2_incremental/smt_terms.h

```diff
@@ -0,0 +1,126 @@
+// Author: Diffblue Ltd.
+
+#ifndef CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_TERMS_H
+#define CPROVER_SOLVERS_SMT2_INCREMENTAL_SMT_TERMS_H
+
+#include <solvers/smt2_incremental/smt_sorts.h>
+#include <util/irep.h>
+
+#include <functional>
+
+class BigInt;
+using mp_integer = BigInt;
+
+class smt_sortt;
+class smt_term_const_downcast_visitort;
+
+class smt_termt : protected irept, private smt_sortt::storert<smt_termt>
```

Comment:
Because this gives us more compile-time checking of the soundness of the conversion and better guarantees about the separation of concerns. This applies both in terms of the conversion from `exprt` to `smt_termt` and the conversion from `smt_termt` to string or other representation understood by the solver. We are only currently planning a conversion to SMT2 strings, but by keeping the conversion to strings separate it would theoretically allow for more easily adding an additional conversion algorithm such as to `smt-switch` data structures to be added on top as an additional run time option. Without the new data structures, the conversion from `exprt` to `smt_termt` could "leak" unconverted `exprt` classes which can only be caught at run time. The way the visitor is defined will also mean that missing conversions from `smt_termt` to string would also cause a compile error. There has to be at least a placeholder method in place with the `UNIMPLEMENTED` macro. 

Defining the data structure will give us a clean separation between the conversion from `exprt` to `smt_termt` and `smt_termt` to string. It will allow for separately unit testing the two conversions. A new data structure only for use with `smt2` has the same advantages that having separate data structures from the language front ends such as for `Java` and `C` vs the internal `goto_program` representation would have.

Reply:
I think you can achieve much of the compile time checking at a much lower cost of having something like...

```
class smt_sortt : public typet {};
class smt_termt : public exprtt {};

class smt_translator {
public :
  virtual smt_sortt type_to_sort(const typet &t) = 0;
  virtual smt_termt expr_to_term(const exprt &e) = 0;
}
```
If this is the only way of producing these two types and the rest of the code is `const`-clean and the critical functions take `smt_sortt` and `smt_termt` then it will be very rare to have failed-to-convert bug.

I would suggest an object because these transformations tend to wind up being stateful no matter how good everyone's intent is.  They will likely also need to be `stackable` in some sense.  As soon as you have a couple of options on how to translate each theory / type (IIRC this was on your list of goals?) then you will have a combinatorial explosion of normal forms, so it is not so easy to use types* to enforce normal forms.

Using the visitor to do the actual conversion or caching converted things is smart idea and it is good to see it early integrated into the design here.

*.  I could be wrong, if I am I would love to know because the handling of goto-programs has exactly this issue.

Reply:
I am not in favour of publicly inheriting from `typet`, for the following reasons -

- It would support down casting from `typet` to `smt_sortt` and require checks to be written for these casts.
- It would add public methods to `smt_sortt` such as `subtype()` and `source_location()` which don't necessarily make sense in the context of a sort.
- It would add public access to member functions from `irept`, which would allow direct manipulation of the `irept` data structures through weakly typed member functions such as `move_to_sub`.
- It would add public access to `irept`'s `is_nil` member function, whereas I would prefer to make the classes in this hierarchy non-nullable/nillable. That way `optionalt` can be added on top of these classes only in the places where nullability is a desirable property.

On the point of the translator - Yes, we will need functionality to convert from `exprt` to `termt`. The choice of whether to use member functions or free functions with configuration/state struct parameters to do the conversions is mostly a stylistic choice in my opinion, rather than a crucial difference. This is unless we really need virtual inheritance and multiple different translators. I have found free functions to be a little easier to unit test in some circumstances, which is a point in their favour. I do think it would improve readability to store configuration and state separately assuming we end up needing both though. That way the configuration can be kept const and the state mutable.